platform/macos/os_macos.mm

```diff
@@ -758,6 +760,113 @@
 	}
 }
 
+NSData *read_data(NSFileHandle *p_fh) {
+	NSData *data = nil;
+	NSString *error_msg = nil;
+#if defined(__x86_64__)
+	if (@available(macOS 10.15, *)) {
+#endif
+		NSError *err = nil;
+		data = [p_fh readDataToEndOfFileAndReturnError:&err];
+		if (data == nil) {
+			error_msg = err.localizedDescription;
+		}
+#if defined(__x86_64__)
+	} else {
+		@try {
+			data = [p_fh readDataToEndOfFile];
+		} @catch (NSException *exception) {
+			error_msg = exception.reason;
+		}
+	}
+#endif
+	if (data == nil) {
+		ERR_PRINT(vformat("read failed: %s", conv::to_string(error_msg)));
+	}
+	return data;
+}
+
+NSString *resolve_path(NSString *p_path) {
+	NSTask *task = [NSTask new];
+	task.executableURL = [NSURL fileURLWithPath:@"/usr/bin/which"];
+	task.arguments = @[ p_path ];
+
+	NSPipe *pipe = [NSPipe pipe];
+	task.standardOutput = pipe;
+
+	NSError *err = nil;
+	if ([task launchAndReturnError:&err] == NO) {
+		ERR_PRINT(vformat("launch failed: %s", conv::to_string(err.localizedDescription)));
+		return p_path;
+	}
+	[task waitUntilExit];
+
+	if (task.terminationStatus != 0) {
+		ERR_PRINT(vformat("command failed: status(%d), reason(%d)", task.terminationStatus, task.terminationReason));
+		return p_path;
+	}
+
+	NSData *data = read_data(pipe.fileHandleForReading);
+	if (data == nil) {
+		return p_path;
+	}
+	NSString *cmd_path = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+	cmd_path = [cmd_path stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceAndNewlineCharacterSet];
+	return cmd_path;
+}
+
+Error OS_MacOS::execute(const String &p_path, const List<String> &p_arguments, String *r_pipe, int *r_exitcode,
+		bool read_stderr, Mutex *p_pipe_mutex, bool p_open_console) {
+	NSString *resolved_path = resolve_path(conv::to_nsstring(p_path));
+
+	NSTask *task = [NSTask new];
+	task.executableURL = [NSURL fileURLWithPath:resolved_path];
+	NSMutableArray<NSString *> *args = [NSMutableArray arrayWithCapacity:p_arguments.size()];
+	for (const String &arg : p_arguments) {
+		[args addObject:conv::to_nsstring(arg)];
+	}
+
+	task.arguments = args;
+
+	NSPipe *output = nil;
+	if (r_pipe) {
+		output = [NSPipe pipe];
+		task.standardOutput = output;
+		if (read_stderr) {
+			task.standardError = output;
+		}
+	}
+
+	NSError *err = nil;
+	if ([task launchAndReturnError:&err] == NO) {
+		ERR_PRINT(vformat("launch failed: %s", conv::to_string(err.localizedDescription)));
+		return FAILED;
+	}
+	[task waitUntilExit];
```

Comment:
Generic Unix implementation is not waiting for completion, but instead reading from the pipe as data arrives. Same with the Windows implementation. So it's not identical behavior.

Reply:
> It is waiting, as that is what [pclose](https://www.man7.org/linux/man-pages/man3/pclose.3p.html) does

It's waiting only after reading is done, not before it. So if it's called from a thread, another thread can get partially read pipe data while the process is still running. But I'm not sure if it's actually used anywhere. Exposed GDScript wrapper only return when pipe is closed, so it can be used only internally.